File: backend/app/schemas/entity.py
import uuid
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, EmailStr, Field

from app.models.entity import EntityType, KYCStatus, RiskLevel


class EntityFilters(BaseModel):
    entity_type: EntityType | None = None
    risk_level: RiskLevel | None = None
    kyc_status: KYCStatus | None = None
    country_code: str | None = None
    is_watchlisted: bool | None = None
    search: str | None = None


class EntityCreate(BaseModel):
    external_id: str = Field(min_length=1, max_length=255)
    entity_type: EntityType
    name: str = Field(min_length=1, max_length=255)
    email: EmailStr | None = None
    phone: str | None = Field(default=None, max_length=50)
    country_code: str | None = Field(default=None, max_length=2)


class EntityUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    email: EmailStr | None = None
    phone: str | None = Field(default=None, max_length=50)
    country_code: str | None = Field(default=None, max_length=2)
    kyc_status: KYCStatus | None = None


class EntityResponse(BaseModel):
    id: uuid.UUID
    external_id: str
    entity_type: EntityType
    name: str
    email: str | None
    phone: str | None
    country_code: str | None
    risk_score: Decimal
    risk_level: RiskLevel
    is_watchlisted: bool
    kyc_status: KYCStatus
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class EntityListResponse(BaseModel):
    total: int
    page: int
    page_size: int
    total_pages: int
    items: list[EntityResponse]


class EntityDetailResponse(EntityResponse):
    """Extended entity response with aggregated stats for the Entity 360 view."""
    transaction_count: int = 0
    total_transaction_amount: Decimal = Decimal("0")
    alert_count: int = 0
    open_case_count: int = 0

File: frontend/src/types/entity.ts
export type EntityType = "individual" | "business" | "merchant";
export type KYCStatus = "pending" | "verified" | "rejected" | "expired";
export type RiskLevel = "low" | "medium" | "high" | "critical";

export interface Entity {
  id: string;
  external_id: string;
  entity_type: EntityType;
  name: string;
  email: string | null;
  phone: string | null;
  country_code: string | null;
  risk_score: number;
  risk_level: RiskLevel;
  is_watchlisted: boolean;
  kyc_status: KYCStatus;
  created_at: string;
  updated_at: string;
}

export interface EntityDetail extends Entity {
  transaction_count: number;
  total_transaction_amount: number;
  alert_count: number;
  open_case_count: number;
}

export interface EntityFilters {
  entity_type?: string;
  risk_level?: string;
  kyc_status?: string;
  country_code?: string;
  is_watchlisted?: boolean;
  search?: string;
}

export interface EntityListResponse {
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
  items: Entity[];
}
